rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+# Continuous Deployment
+
+## Summary
+
+In [RFC 70](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-070-path-towards-continuous-deployment-cd.md#proposal) we committed to fully automatic deployments. The following provides a framework for how we will actually achieve this for all of our existing apps.
+
+Decision: in order to safely enable automatic deployments for each app, we will specify minimum criteria that must first be met. See the criteria described [below](#safety-concerns).
+
+An audit of our apps against these criteria can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+## Problem
+
+Merging a PR for an application repo [results in a new "release" tag](https://github.com/alphagov/govuk-jenkinslib/blob/30e9392d31b98e5c251ea471d793761748eea1e8/vars/govuk.groovy#L123). We have a [Release](https://github.com/alphagov/release) app to monitor these. Every new release tag [is automatically deployed to our Integration environment](https://github.com/alphagov/govuk-jenkinslib/blob/30e9392d31b98e5c251ea471d793761748eea1e8/vars/govuk.groovy#L134).
+
+This is where the automation ends. In order for a release (or series of releases) to be deployed to the GOV.UK Production environment, a developer with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) must:
+
+- Manually check the release is "OK" in Integration.
+- Manually deploy the release to Staging.
+- Manually check the release is "OK" in Staging.
+- Manually deploy the release to Production.
+
+Having these manual steps has lead to a number of problems.
+
+### We freely merge PRs with breaking changes
```

Comment:
Although your logic sounds accurate @sengi I think the deployment related problem of "we merge breaking changes" is that we've not fully established that merging means that the code is safe to deploy. This is due to expectations that any problems with the merge would be caught either during testing as part of a manual deploy or by manual testing on integration post merge.

Reply:
I agree that's an important issue. I think it needs to be made clearer as to whether it's in or out of scope of the RFC though.

Reply:
This is an interesting discussion. I've always thought of "enabling CD" as being synonymous with "enabling CD safely". I don't think it's possible to have a discussion about RFC and not consider safety. 

We can break "safety" down into:

- Poor testing i.e. a poor definition of "OK". This is addressed in the section dedicated to safety criteria i.e. we _will_ be increasing our automated coverage and reducing our reliance on manual testing.

- Unsafe habits (other than poor testing 😆) i.e. freely merging stuff that breaks. The only decision we're making here is whether to enable CD. In the "being more cautious" consequence, I'm trying to argue that we will naturally address this problem - we can't force good habits on people, but we can make bad habits less appealing!

Of the two, I think the latter is the one that could be seen as out-of-scope. However, I really do believe that enabling CD will help alter these habits (and has already), and I'm open to other ideas to push that.

Thanks for the very clear suggestion, by the way! I'm open to changing things. How about we have a chat about this? (Long textual discussion can get a bit demoralising after a while.)

Reply:
No worries - more than happy to have a quick chat instead, might be easier. I went and muddied the waters by bringing the scope question into it. My original point was meant to be about document structure and clarity rather than scope.

Reply:
Me and @sengi have discussed this and agreed it's superseded by the discussion we're having on https://github.com/alphagov/govuk-rfcs/pull/128#discussion_r510868062.